src/main/java/frc/team2412/robot/subsystem/TargetLocalizer.java

```diff
@@ -0,0 +1,74 @@
+package frc.team2412.robot.subsystem;
+
+import org.frcteam2910.common.math.Rotation2;
+
+import static frc.team2412.robot.subsystem.TargetLocalizer.LocalizerConstants.*;
+
+public class TargetLocalizer {
+    public static class LocalizerConstants {
+        public static final double TURRET_OFFSET = 0;
+        // TODO tune these more
+        public static final double TURRET_LATERAL_FF = 4.8, TURRET_ANGULAR_FF = 4.8, TURRET_DEPTH_FF = 0,
+                LATERAL_DEPTH_COMPENSATION = 0;
+    }
+
+    private final DrivebaseSubsystem drivebaseSubsystem;
+    private final ShooterSubsystem shooterSubsystem;
+    private final ShooterVisionSubsystem shooterVisionSubsystem;
+
+    public TargetLocalizer(DrivebaseSubsystem drivebase, ShooterSubsystem shooter, ShooterVisionSubsystem vision) {
+        drivebaseSubsystem = drivebase;
+        shooterSubsystem = shooter;
+        shooterVisionSubsystem = vision;
+    }
+
+    public double getDistance() {
+        return hasTarget() ? shooterVisionSubsystem.getDistance() + shooterSubsystem.getDistanceBias() : 0;
+    }
+
+    public double getAdjustedDistance() {
+        return getDistance() + distanceAdjustment();
+    }
+
+    public double distanceAdjustment() {
+        return (getDepthVelocity() * TURRET_DEPTH_FF) / getVoltage();
+    }
+
+    public double getPitch() {
+        return shooterVisionSubsystem.getAdjustedPitch();
+    }
+
+    public boolean hasTarget() {
+        return shooterVisionSubsystem.hasTarget();
+    }
+
+    public double getYaw() {
+        // return 0;
+        return shooterVisionSubsystem.getYaw() + shooterSubsystem.getTurretAngleBias();
+    }
+
+    public double getLateralVelocity() {
+        return drivebaseSubsystem.getVelocity()
+                .rotateBy(Rotation2.fromDegrees(TURRET_OFFSET + shooterSubsystem.getTurretAngle())).x;
+    }
+
+    public double getDepthVelocity() {
+        return drivebaseSubsystem.getVelocity()
+                .rotateBy(Rotation2.fromDegrees(TURRET_OFFSET + shooterSubsystem.getTurretAngle())).y;
+    }
+
+    public double getAngularVelocity() {
+        System.out.println(drivebaseSubsystem.getAngularVelocity());
+        return drivebaseSubsystem.getAngularVelocity();
+    }
+
+    public double yawAdjustment() {
+        return (getLateralVelocity() * TURRET_LATERAL_FF + getAngularVelocity() * TURRET_ANGULAR_FF) / getVoltage()
+                + getDistance() * LATERAL_DEPTH_COMPENSATION;
```

Comment:
As distance increases, the adjustment should decrease, not increase (because smaller changes in angle will have the same impact on where the shot lands).

Reply:
can make depth comp negative ez

Reply:
Also note that if there is no movement (getLaterVelocity() is 0 and getAngularVelocity() is 0), we want the yaw adjustment to be 0, but as it currently is, it would return non-zero because getDistance() would be non-zero (and LATERAL_DEPTH_COMPENSATION would be non-zero if we're going to do anything).
Honestly, I think we should switch to calculating the position of the hub relative to the turret and then transform + rotate by the robot's velocity multiplied by a certain length of time (an estimate of how long from beginning of shot to ball landing).

Reply:
effort, it will work okayish when we tune